larsclock.man
.TH larsclock 1x
.SH NAME
larsclock \- Status bar clock for larswm
.SH SYNOPSIS
.B larsclock
[
.B \-display
.I dpy
]
[
.B \-format
.I fmt
]
[
.B \-update
.I seconds
]
.SH DESCRIPTION
.I larsclock
is a minimal status bar clock application for the larswm window manager.
.SH OPTIONS
The following command line options are supported:
.TP 18
.BI \-display " dpy"
lets you specify which display you want
.I larsclock
to open.
.TP 18
.BI \-format " fmt"
lets you format how the time is displayed. The default format is
"%c". See strftime(3) for details.
.TP 18
.BI \-update " seconds"
lets you specify the  frequency in seconds at which larsclock
should update status bar.
.SH SEE ALSO
larsmenu(1x), larsremote(1x), larswm(1x), strftime(3x)
.SH AUTHORS
.I larsclock
was created by Lars Bernhardsson <lab@fnurt.net>.


ChangeLog
1.0 - Jun  3, 2000:   1. Initial version. Added tiling to 9wm.

1.1 - Jun  5, 2000:   1. Added support for max size hints.
                      2. Also, Things with width_inc and height_inc
                         should look better now.

1.2 - Jun  5, 2000:   1. Incorporated ideas/code from Benjamin Drieu's
                         w9wm project, so now it has virtual desktops,
                         currently hardcoded at four.

1.3 - Jun  6, 2000:   1. Fixed a bug with PMaxSize windows that set hints
                         after the window has been mapped.
                      2. Fixed some other small bugs.

1.4 - Jun  6, 2000:   1. Added support for Prev. Desktop.
                      2. When toggling Auto ReTile, it toggles for the
                         current desktop only! Now you can have some
                         desktops auto tile, and some behave more
                         traditional.
                      3. Added support for Button4 and Button5.
                         Button4 switches to previous desktop, Button5
                         to next. If you have a wheel mouse, put:
                         ZAxisMapping 4 5 in your XF86Config to use
                         wheel up and down to switch desktops.
                      4. Added the -rv option. Reverses black/white in
                         all window manager menus/windows.

1.4.1 - Jun  7, 2000: 1. Fixed the way transient windows are raised,
                         now nested transients  work properly.
                      2. renamed setstate() to wmsetstate() to avoid
                         clashing with the function of the same name
                         in stdlib.h.
                      3. Fixed a bug in how transient windows are hidden
                         and restored when moving between desktops.

1.5 - Jun  8, 2000:   1. Added support for 'tools', applications you
                         always want visible on the screen and not
                         included in the retiling process.
                         See -toolspace and -toolclass command line
                         options.
                      2. Fixed some minor cosmetic bugs. Using all
                         screen space optimally.

2.0 - Jun  9, 2000:   1. Added support for multi-screen displays.
                         You can change desktop, default number of
                         windows in left track, auto tiling on/off
                         individually for each screen.
                         You can also pick which screen you want
                         toolspace reserved on.

2.1 - Jun 15, 2000:   1. Added support for focusing a window without
                         retiling everything, even if auto tiling
                         is on for that screen/desktop.
                         Just Ctrl-LeftClick on an inactive window.
                      2. Added an option to always act as if window
                         selects were done with Control held down.
                      3. Added an option to balance tracks. This
                         makes sure that both tracks have the same width
                         and number of windows in them. Works nice
                         with SkipFocus for an Oberon effect.

2.2 - Jun 27, 2000:   1. Added the -nouserhints option. Now by default,
                         if the user specifies the position of a window,
                         it will not be included in the tiling process.
                         This can be used instead of -toolclass as a
                         way of not having to specify every kind of
                         window you don't want tiled, or if you want
                         some but not all windows of a class tiled.
                      2. If the argument to -toolspace is a negative
                         number, space is reserved at the top of the
                         screen instead of the bottom.
                      3. Added the -widetools option. This makes the
                         reserved toolspace stretch across the screen,
                         not just across the right track.

3.0 - Jun 28, 2000:   1. Big cleanup. Removed -toolspace, -toolclass
                         -toolscreen and -widetools options.
                         Now tools are looked for at each retiling
                         event, and anything that has had its
                         position specified by the user is considered
                         a tool.
                         Where to set aside toolspace (top or bottom,
                         right track only or full width) is calculated.
                         When a tool pops up, it might end up on top
                         of a tiled window, it will be properly tiled
                         when the next retiling event occurs.

3.1 - Jun 30, 2000:   1. Rearranged all menus. :)
                         Here's how the menus work, they can be popped
                         up anywhere on the root window, or over an
                         inactive window.
                         Mouse Button    Function/Menu
                         ------------------------------------------------
                         Right           Make clicked window active.
                         Control+Right   Active, without retiling.

                         Middle          Next Desktop.
                         Shift+Middle    List all Desktops.
                         Control+Middle  Previous Desktop.

                         Left            Main menu, same as 9wm.
                         Shift+Left      Hidden Windows.
                         Control+Left    larswm functions.

                         Shift+Control Left or Right shows version info.
                         ------------------------------------------------
                      2. Cleaned up the usage of pointers to active

larsclock.c
/* Copyright (c) 2000 Lars Bernhardsson, see README for licence details */

/*
 * Minimal clock app for larswm. No fork()/exec() is done, it updates
 * the status bar by setting the LARSWM_BARTEXT atom every second.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <time.h>

#ifdef USE_LOCALE
#include <locale.h>
#endif

#include <X11/X.h>
#include <X11/Xos.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>

#define CLOCK_LENGTH 100

int
main (int argc, char **argv)
{
  int i;
  int update = 0;
  time_t t;
  char *slist;
  char *format = "%c";
  char *display_string = "";
  Display *dpy;
  Atom bartext_larswm;
  XTextProperty pr;

#ifdef USE_LOCALE
  setlocale (LC_ALL, "");
#endif

  for (i = 1; i < argc; i++)
    {
      if (strcmp (argv[i], "-display") == 0 && i + 1 < argc)
	{
	  display_string = argv[++i];
	}
      else if (strcmp (argv[i], "-format") == 0 && i + 1 < argc)
	{
	  format = argv[++i];
	}
      else if (strcmp (argv[i], "-update") == 0 && i + 1 < argc)
	{
	  update = atoi(argv[++i]);
	}
      else
	{
	  fprintf (stderr, "usage: %s [-display display] [-format format] [-update seconds]\n",
		   argv[0]);
	  exit (1);
	}
    }

  dpy = XOpenDisplay (display_string);

  if (dpy == 0)
    {
      perror ("XOpenDisplay");
      exit (1);
    }

  if (update == 0)
    {
      update = 1;
    }
  
  bartext_larswm = XInternAtom (dpy, "LARSWM_BARTEXT", False);
  slist = malloc (CLOCK_LENGTH);

  while (1)
    {
      time (&t);
      strftime (slist, CLOCK_LENGTH, format, localtime (&t));

      XStringListToTextProperty (&slist, 1, &pr);
      XSetTextProperty (dpy, DefaultRootWindow (dpy), &pr, bartext_larswm);
      XFree (pr.value);
      XSync (dpy, False);

      sleep (update);
    }
}
